docs/developer-docs/backend/rust/1-infrastructure.md

```diff
@@ -0,0 +1,204 @@
+
+## Observability
+Metrics can be used to gain insight into a wide range of information regarding your canister's production services. This data is important to learn about your canister's statistics and productivity. 
+
+### Exposing canister metrics
+
+#### Approach 1: Expose a query call that returns a data structure containing your canister's metrics. 
+If this data is not intended to be public, this query can be configured to be rejected based on the caller's principal. This approach provides an response that is structured and easy to parse. 
+
+```
+pub struct MyMetrics {   
+  pub stable_memory_size: u32,   
+  pub allocated_bytes: u32,   
+  pub my_user_map_size: u64,   
+  pub last_upgraded_ts: u64, 
+}
+#[query] 
+fn metrics() -> MyMetrics {   
+  check_acl();   
+  MyMetrics {     
+    // ...   
+  } 
+}
+```
+
+#### Approach 2: Expose the canister's metrics in a format that your monitoring system can ingest through the canister's HTTP gateway. 
+
+For text-based exposition formats, the following example can be used:
+
+```
+fn http_request(req: HttpRequest) -> HttpResponse {   
+  match path(&req) {     
+    "/metrics" => HttpResponse {         
+        status_code: 200,         
+        body: format!(r#"stable_memory_bytes {}
+                         allocated_bytes {}  
+                         registered_users_total {}"#, 
+                      stable_memory_bytes, allocated_bytes, num_users),         
+        // ...     
+    }   
+  } 
+}
+```
+
+#### Important metric data to watch
+- The size of the canister's stable memory.
+- The size of the canister's internal data structures
+- The sizes of objects allocated within the heap.
+- The date and time the canister was last upgraded.
+
+## Globally mutable states
+
+By design, canisters on the IC are structured in a way that forces developers to use a global mutable state. However, Rust's design makes it difficult to global mutable variables. This results in Rust developers needing to choose a method of code organization that takes the IC's design into consideration. This guide will cover a few of those code organization options. 
+
+### Using `thread_local!` with `Cell/RefCell` for state variables
+
+Using `thread_local!` with `Cell/RefCell` is the safest option to avoid issues with asynchrony and memory corruption. 
+
+The following is an example of how `thread_local!` can be used:
+
+```
+thread_local! {
+    static NEXT_USER_ID: Cell<u64> = Cell::new(0);
+    static ACTIVE_USERS: RefCell<UserMap> = RefCell::new(UserMap::new());
+}
+```
+
+### Using `let state = ic_cdk::storage::get_mut<MyState>`
+
+The IC [Rust CDK](https://github.com/dfinity/cdk-rs) provides storage abstraction that allows you to get a mutable reference indexed by a type through the `get_mut<MyState>` method. This allows you to obtain multiple non-exclusive mutable references to the same object, which should be used with caution. 
+
+
```

Comment:
I believe `ic_cdk::storage::get_mut` is retired in ic_cdk 0.5; we can remove this section.